### ISSUE TITLE
CRD is returning the whole graph

### ISSUE THREAD

**Issue by dgleich** [kfoynt/LocalGraphClustering#62]
For the following example, I get the entire set of nodes of the back from CRD. This seems like a bug in that we shouldn't ever get the entire graph back (all but one node, sure... but, not the entire graph...) 
```
S = lgc.flow_clustering(helper.lgc_data("ASTRAL"),[398],method="crd", U = 3,h = 10, w = 2)
g.set_scores(S[0])

{'cond': 1,
 'cut': 0.0,
 'edgeseff': 0.0,
 'edgestrue': 314428.0,
 'isop': 1,
 'sizeeff': 0,
 'sizetrue': 1049,
 'voleff': 0.0,
 'voltrue': 314428.0}
```

**Comment by MengLiuPurdue** [comment]
@dgleich Not sure what happened, but I am getting different results.

{'isop': 639.1168815418222,
 'cond': 0.5369737740203705,
 'edgeseff': 191783.85685539246,
 'edgestrue': 191783.85685539246,
 'sizeeff': 348,
 'sizetrue': 348,
 'voleff': 414196.53163194656,
 'cut': 222412.6747765541,
 'voltrue': 414196.53163194656}

**Comment by dgleich** [comment]
That's even worse :) Have you pulled the latest code?

**Comment by MengLiuPurdue** [comment]
Yes, pulled the latest code and recompiled the C++.

**Comment by MengLiuPurdue** [comment]
@dgleich Yes, I just tried and still got the same results. I noticed that in your code, "g" is actually not defined. Is it possible that "g" is defined mistakenly by some other graph instead of "ASTRAL"?

**Comment by dgleich** [comment]
I originally got these out of the NCP, so it must have been the same graph.
Will see if I can debug further...

David

**Comment by MengLiuPurdue** [comment]
@dgleich One issue I noticed in your results is that "ASTRAL" is a weighted graph, but your "voltrue" is the total number of edges instead of the sum of all degrees. So your "g" must be defined as an unweighted graph somewhere, but a weighted graph is passed into "crd".

**Comment by dgleich** [comment]
Call discard_weights on the graph after loading ...

**Comment by dgleich** [comment]
This is getting stranger and stranger...

There is definitely a bug with CRD. A bad one, from what it appears. 

If I run a crd ncp. 
```
ncp = lgc.NCPData(g)
ncp.crd(nthreads=10)
```

Then some of the sets come back full. If I reload the graph, and re-run the CRD from these nodes, I see that they also come back full. But then if I restart the kernel, and repeat the same CRD run from these nodes, they come back not full.

So the order is:

1. Run NCP
2. Find a node that will return a set of vertices for the entire graph.
3. Reload the graph.
4. Run that same CRD on the newly loaded graph.
5. See that the set comes back full.

--

Is there some why CRD could be not-freeing memory that would case this type of effect?

**Comment by dgleich** [comment]
```
import localgraphclustering as lgc
import sys
import os
sys.path.insert(0, os.path.join("..", "LocalGraphClustering", "notebooks"))
import helper

g = helper.lgc_data("ASTRAL")
g.discard_weights()
n = g._num_vertices
for i in range(100):
    R = [i]
    R.extend(g.neighbors(i))
    S = lgc.flow_clustering(g,R,method="crd", U = 3,h = 10, w = 2)
    if len(S[0]) == n:
        print("error on node ", i)
        break
```

This seems to reliably elicit the problem.

**Comment by dgleich** [comment]
However, it will be a random node! I've seen it on node 7, node 20, 21, 25, sometimes it'll work...

this argues that there is some issue with un-initialized memory... or we are somehow over write info?

**Comment by dgleich** [comment]
Run it a few times on unimodular. There is something VERY strange going on.

David

**Comment by MengLiuPurdue** [comment]
Sure, my unimodular has some C++ libraries missing. I will need to install them first. which might take some time.

**Comment by dgleich** [comment]
What do you mean by this? Did you install a custom compiler?

**Comment by MengLiuPurdue** [comment]
Not sure, this is what I get when I try to compile it.

g++ -g -std=c++11 -Wall -Wshadow -fPIC   -c -o ../sweepcut.o ../sweepcut.cpp
In file included from /usr/include/stdio.h:27:0,
                 from ../sweepcut.cpp:34:
/usr/include/features.h:367:12: fatal error: sys/cdefs.h: No such file or directory
 #  include <sys/cdefs.h>
            ^~~~~~~~~~~~~
compilation terminated.
<builtin>: recipe for target '../sweepcut.o' failed
make: *** [../sweepcut.o] Error 1

**Comment by dgleich** [comment]
The issue is in the capacity_releasing_diffusion.cpp file...

The array cond_best_array can contain NaN values. This likely represents bad initialized memory...

**Comment by dgleich** [comment]
Fixed by #63
